File: sqlite-android-studio-plugin/src/main/java/com.alecstrong.sqlite.android/psi/SqliteElement.java
package com.alecstrong.sqlite.android.psi;

import com.alecstrong.sqlite.android.util.SqlitePsiUtils;
import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiNamedElement;
import com.intellij.psi.tree.IElementType;
import com.intellij.util.IncorrectOperationException;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;

public abstract class SqliteElement extends ASTWrapperPsiElement implements PsiNamedElement {
  protected String name = null; // an override to input text ID

  public SqliteElement(@NotNull final ASTNode node) {
    super(node);
  }

  @Override
  public String getName() {
    if ( name!=null ) return name;
    IdentifierElement id = getId();
    if ( id!=null ) {
      return id.getText();
    }
    return "unknown-name";
  }

  public abstract IdentifierElement getId();

  @Override
  public PsiElement setName(@NonNls @NotNull String name) throws IncorrectOperationException {
		/*
		From doc: "Creating a fully correct AST node from scratch is
		          quite difficult. Thus, surprisingly, the easiest way to
		          get the replacement node is to create a dummy file in the
		          custom language so that it would contain the necessary
		          node in its parse tree, build the parse tree and
		          extract the necessary node from it.
		 */
    //		System.out.println("rename "+this+" to "+name);
    IdentifierElement id = getId();
    id.replace(SqlitePsiUtils.createLeafFromText(getProject(),
        getContext(),
        name, getRuleRefType()));
    this.name = name;
    return this;
  }

  public abstract IElementType getRuleRefType();

  @Override
  public void subtreeChanged() {
    super.subtreeChanged();
    name = null;
  }

  @Override
  public int getTextOffset() {
    IdentifierElement id = getId();
    if ( id!=null ) return id.getTextOffset();
    return super.getTextOffset();
  }
}

File: sqlite-android-studio-plugin/src/main/kotlin/com/alecstrong/sqlite/android/psi/SqliteElement.kt
package com.alecstrong.sqlite.android.psi

import com.alecstrong.sqlite.android.SQLiteLexer
import com.alecstrong.sqlite.android.lang.SqliteTokenTypes
import com.alecstrong.sqlite.android.util.SqlitePsiUtils
import com.intellij.lang.ASTNode
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiNamedElement
import com.intellij.psi.tree.IElementType
import com.intellij.psi.util.PsiTreeUtil

sealed class SqliteElement(node: ASTNode) : ASTWrapperPsiElement(node), PsiNamedElement {
  protected var hardcodedName: String? = null

  abstract protected val ruleRefType: IElementType

  val id: IdentifierElement?
    get() = PsiTreeUtil.findChildOfType(this, IdentifierElement::class.java)

  override fun getTextOffset() = id?.textOffset ?: super.getTextOffset()
  override fun getName() = hardcodedName ?: id?.text ?: "unknown-name"
  override fun setName(name: String): PsiElement {
    id?.replace(SqlitePsiUtils.createLeafFromText(project, context, name, ruleRefType))
    hardcodedName = name
    return this
  }

  override fun subtreeChanged() {
    super.subtreeChanged()
    hardcodedName = null
  }

  internal class TableNameElement(node: ASTNode) : SqliteElement(node) {
    override val ruleRefType = SqliteTokenTypes.TOKEN_ELEMENT_TYPES[SQLiteLexer.IDENTIFIER]

    fun isSameTable(other: TableNameElement?) = id?.name?.equals(other?.id?.name) ?: false
  }

  internal class ColumnNameElement(node: ASTNode) : SqliteElement(node) {
    override val ruleRefType = SqliteTokenTypes.TOKEN_ELEMENT_TYPES[SQLiteLexer.IDENTIFIER]
  }
}
